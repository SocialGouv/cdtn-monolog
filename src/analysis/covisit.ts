import { IDataFrame } from "data-forge";

import * as datasetUtil from "../reader/dataset";
import { urlToPath } from "../reader/readerUtil";
import { CovisiteReport } from "./reports";

const reportType = "covisit";

// number of covisited links to be added to each content report
const LINK_LIMIT = 6;

// minimum covisits to consider as actual link
// 5 means : a minimum of 5 unique visits where two contents were viewed
const MIN_OCC = 5;
const analyse = (
  dataset: IDataFrame,
  minOcc: number = MIN_OCC,
  linkLimit: number = LINK_LIMIT,
  reportId: string = new Date().getTime().toString()
): CovisiteReport[] => {
  // FIXME avoid using to array
  const visits = datasetUtil.getVisits(dataset).toArray();

  const uniqueViews = visits.map((v) =>
    datasetUtil.toUniqueViews(v).deflate((a: any) => a.url)
  );

  // unique key to represent link between a & b
  const toKey = (a: string, b: string) => (a < b ? `${a}__${b}` : `${b}__${a}`);

  const graph = new Map();

  uniqueViews.forEach((vv) => {
    const visitViews = vv.toArray();
    if (visitViews.length > 1) {
      // visitViews.map((v) => console.log(v.toString()));
      // get all pairs of contents A/B
      const covisits = visitViews.reduce(
        (acc: string[][], a: string, i: number) => {
          acc.push(...visitViews.slice(i + 1).map((b: string) => [a, b]));
          return acc;
        },
        []
      );

      // TODO use proper edge type rather than array
      covisits.forEach(([a, b]: string[]) => {
        const key = toKey(a, b);
        if (!graph.has(key)) {
          graph.set(key, { count: 1, nodes: [a, b] });
        } else {
          graph.get(key).count++;
        }
      });
    }
  });

  const contents = new Map<string, Map<string, number>>();

  graph.forEach((edge) => {
    const [a, b] = edge.nodes;

    const addToContent = (x: string, y: string) => {
      if (!contents.has(x)) {
        contents.set(x, new Map());
      }
      contents.get(x)?.set(y, edge.count);
    };

    // only keep if count is more than MIN_OCC
    if (edge.count >= minOcc) {
      addToContent(a, b);
      addToContent(b, a);
    }
  });

  const sortedContents = new Map<string, Map<string, number>>();
  // sort contents by count
  contents.forEach((counts, url) => {
    const sorted = new Map([...counts.entries()].sort((a, b) => b[1] - a[1]));
    sortedContents.set(url, sorted);
  });

  const docs: CovisiteReport[] = [];

  sortedContents.forEach((counts, url) => {
    const content = urlToPath(url);

    const links = Array.from(counts)
      .slice(0, linkLimit)
      .map(([link, count]) => {
        return {
          count,
          link: urlToPath(link),
        };
      });
    docs.push({ content, links, reportId, reportType });
  });

  return docs;
};

export { analyse, reportType };
