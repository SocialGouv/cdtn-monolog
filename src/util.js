export const SERVICE_URL = "https://code.travail.gouv.fr/";

export const urlToPath = (url) => {
  if (url.startsWith(SERVICE_URL)) {
    return url.slice(SERVICE_URL.length);
  } else {
    return url;
  }
};

export const actionTypes = {
  visitType: "visit_content",
  selectSuggestionType: "select_suggestion",
};
