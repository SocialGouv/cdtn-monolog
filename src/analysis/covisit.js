import * as datasetUtil from "../dataset";
import * as util from "../util";

const reportType = "covisit";

// number of covisited links to be added to each content report
const LINK_LIMIT = 6;

// minimum covisits to consider as actual link
// 5 means : a minimum of 5 unique visits where two contents were viewed
const MIN_OCC = 5;

const analyse = (dataset, minOcc = MIN_OCC, linkLimit = LINK_LIMIT) => {
  // FIXME avoid using to array
  const visits = datasetUtil.getVisits(dataset).toArray();

  const uniqueViews = visits.map((v) =>
    datasetUtil.toUniqueViews(v).deflate((a) => a.url)
  );

  // unique key to represent link between a & b
  const toKey = (a, b) => (a < b ? `${a}__${b}` : `${b}__${a}`);

  const graph = new Map();

  uniqueViews.forEach((vv) => {
    const visitViews = vv.toArray();
    if (visitViews.length > 1) {
      // visitViews.map((v) => console.log(v.toString()));
      // get all pairs of contents A/B
      const covisits = visitViews.reduce((acc, a, i) => {
        acc.push(...visitViews.slice(i + 1).map((b) => [a, b]));
        return acc;
      }, []);

      covisits.forEach(([a, b]) => {
        const key = toKey(a, b);
        if (!graph.has(key)) {
          graph.set(key, { nodes: [a, b], count: 1 });
        } else {
          graph.get(key).count++;
        }
      });
    }
  });

  const contents = new Map();

  graph.forEach((edge) => {
    const [a, b] = edge.nodes;

    const addToContent = (x, y) => {
      if (!contents.has(x)) {
        contents.set(x, new Map());
      }
      contents.get(x).set(y, edge.count);
    };

    // only keep if count is more than MIN_OCC
    if (edge.count >= minOcc) {
      addToContent(a, b);
      addToContent(b, a);
    }
  });

  const sortedContents = new Map();
  // sort contents by count
  contents.forEach((counts, url) => {
    const sorted = new Map([...counts.entries()].sort((a, b) => b[1] - a[1]));
    sortedContents.set(url, sorted);
  });

  const docs = [];
  sortedContents.forEach((counts, url) => {
    const content = util.urlToPath(url);

    const links = Array.from(counts)
      .slice(0, linkLimit)
      .map(([link, count]) => {
        return {
          link: util.urlToPath(link),
          count,
        };
      });
    docs.push({ content, links, reportType });
  });

  return docs;
};

// configured function

export { analyse, reportType };
