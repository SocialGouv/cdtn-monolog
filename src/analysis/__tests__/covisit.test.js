// unitish test for covisit
import * as Covisit from "../covisit";
import * as util from "../../util";
import { DataFrame } from "data-forge";

describe("Covisit", () => {
  const toVisitAction = (uvi, url) => ({
    uvi,
    url,
    type: util.actionTypes.visitType,
  });

  const url = (i) => `url${i}`;

  const analyse = (dataset) => Covisit.analyse(dataset, 1, 3);

  // takes tuples and generate data : [1,3] =>  visit id 1 / content viewed 3
  const generate = (a) =>
    new DataFrame(a.map(([uvi, u]) => toVisitAction(uvi, url(u))));

  it("should only count unique visits", () => {
    const data = [
      [1, 1],
      [1, 1],
      [1, 2],
    ];

    const df = generate(data);
    expect(analyse(df)).toMatchSnapshot();
  });

  it("should link covisits", () => {
    const data = [
      [1, 1],
      [1, 2],
      [1, 3],

      [2, 1],
      [2, 3],

      [3, 2],
      [3, 3],

      [4, 1],
      [4, 3],
    ];
    const df = generate(data);
    expect(analyse(df)).toMatchSnapshot();
  });

  it("should respect minimum occurences setting", () => {
    const rawMinOcc = [
      [1, 1],
      [1, 2],
      [2, 1],
      [2, 2],
    ];
    const dataMinOcc = generate(rawMinOcc);

    // try with minoccurence 2
    expect(Covisit.analyse(dataMinOcc, 2, 2)).toMatchSnapshot();

    // try with minoccurence 3
    expect(Covisit.analyse(dataMinOcc, 3, 2)).toStrictEqual([]);
    // add content so it passes limitation
    [
      [3, 1],
      [3, 2],
    ].forEach((a) => rawMinOcc.push(a));
    expect(Covisit.analyse(generate(rawMinOcc), 3, 2)).toMatchSnapshot();
  });

  it("should respect link limit setting", () => {
    const linkLimit = generate([
      [1, 1],
      [1, 2],
      [1, 3],
      [2, 1],
      [2, 2],
      [2, 3],
      [3, 1],
      [3, 2],
      [3, 3],
    ]);

    // try with link limit 2
    expect(Covisit.analyse(linkLimit, 1, 2)).toMatchSnapshot();

    // try with link limit 3
    expect(Covisit.analyse(linkLimit, 1, 1)).toMatchSnapshot();
  });
});
