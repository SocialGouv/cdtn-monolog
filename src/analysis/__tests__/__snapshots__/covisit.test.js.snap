// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Covisit should link covisits 1`] = `
Array [
  Object {
    "content": "url1",
    "links": Array [
      Object {
        "count": 3,
        "link": "url3",
      },
      Object {
        "count": 1,
        "link": "url2",
      },
    ],
    "reportType": "covisit",
  },
  Object {
    "content": "url2",
    "links": Array [
      Object {
        "count": 2,
        "link": "url3",
      },
      Object {
        "count": 1,
        "link": "url1",
      },
    ],
    "reportType": "covisit",
  },
  Object {
    "content": "url3",
    "links": Array [
      Object {
        "count": 3,
        "link": "url1",
      },
      Object {
        "count": 2,
        "link": "url2",
      },
    ],
    "reportType": "covisit",
  },
]
`;

exports[`Covisit should only count unique visits 1`] = `
Array [
  Object {
    "content": "url1",
    "links": Array [
      Object {
        "count": 1,
        "link": "url2",
      },
    ],
    "reportType": "covisit",
  },
  Object {
    "content": "url2",
    "links": Array [
      Object {
        "count": 1,
        "link": "url1",
      },
    ],
    "reportType": "covisit",
  },
]
`;

exports[`Covisit should respect link limit setting 1`] = `
Array [
  Object {
    "content": "url1",
    "links": Array [
      Object {
        "count": 3,
        "link": "url2",
      },
      Object {
        "count": 3,
        "link": "url3",
      },
    ],
    "reportType": "covisit",
  },
  Object {
    "content": "url2",
    "links": Array [
      Object {
        "count": 3,
        "link": "url1",
      },
      Object {
        "count": 3,
        "link": "url3",
      },
    ],
    "reportType": "covisit",
  },
  Object {
    "content": "url3",
    "links": Array [
      Object {
        "count": 3,
        "link": "url1",
      },
      Object {
        "count": 3,
        "link": "url2",
      },
    ],
    "reportType": "covisit",
  },
]
`;

exports[`Covisit should respect link limit setting 2`] = `
Array [
  Object {
    "content": "url1",
    "links": Array [
      Object {
        "count": 3,
        "link": "url2",
      },
    ],
    "reportType": "covisit",
  },
  Object {
    "content": "url2",
    "links": Array [
      Object {
        "count": 3,
        "link": "url1",
      },
    ],
    "reportType": "covisit",
  },
  Object {
    "content": "url3",
    "links": Array [
      Object {
        "count": 3,
        "link": "url1",
      },
    ],
    "reportType": "covisit",
  },
]
`;

exports[`Covisit should respect minimum occurences setting 1`] = `
Array [
  Object {
    "content": "url1",
    "links": Array [
      Object {
        "count": 2,
        "link": "url2",
      },
    ],
    "reportType": "covisit",
  },
  Object {
    "content": "url2",
    "links": Array [
      Object {
        "count": 2,
        "link": "url1",
      },
    ],
    "reportType": "covisit",
  },
]
`;

exports[`Covisit should respect minimum occurences setting 2`] = `
Array [
  Object {
    "content": "url1",
    "links": Array [
      Object {
        "count": 3,
        "link": "url2",
      },
    ],
    "reportType": "covisit",
  },
  Object {
    "content": "url2",
    "links": Array [
      Object {
        "count": 3,
        "link": "url1",
      },
    ],
    "reportType": "covisit",
  },
]
`;
